File: desktop-app/src-tauri/src/commands/signing.rs
use desktop_app::infrastructure::dev_secrets;
use desktop_app::infrastructure::signing;

#[tauri::command]
pub(crate) fn compute_sighash(
    seqno: u64,
    action_hex: String,
) -> Result<signing::SighashResult, String> {
    signing::compute_sighash(seqno, &action_hex)
}

/// Reports whether mnemonic / raw-key signing over IPC is enabled (debug build, or
/// release with ALLOW_DEV_MNEMONIC_SIGNING=1). Lets the UI hide the Mnemonic option when off.
#[tauri::command]
pub(crate) fn dev_mnemonic_signing_enabled() -> bool {
    dev_secrets::dev_mnemonic_signing_ipc_enabled()
}

/// Canonical human-readable SPS-65 signing message for an action — the exact string the
/// device signs. Used by the UI to show what the connected device displays (Trezor renders
/// this text; Ledger renders its SHA-256 as the "Message hash"), instead of the BIP-137
/// sighash which never appears on-device.
#[tauri::command]
pub(crate) fn render_signing_message(seqno: u64, action_hex: String) -> Result<String, String> {
    signing::render_signing_message(seqno, &action_hex)
}

#[tauri::command]
pub(crate) fn verify_threshold(
    public_keys_hex: Vec<String>,
    threshold: u32,
    signatures_hex: Vec<String>,
    sighash_hex: String,
) -> Result<signing::VerifyResult, String> {
    signing::verify_threshold(&public_keys_hex, threshold, &signatures_hex, &sighash_hex)
}

#[tauri::command]
pub(crate) fn sign_action_sighash(
    secret_key_hex: String,
    sighash_hex: String,
) -> Result<signing::SignatureResult, String> {
    dev_secrets::ensure_dev_mnemonic_signing_allowed()?;
    signing::sign_sighash(&secret_key_hex, &sighash_hex)
}

#[tauri::command]
pub(crate) fn list_mnemonic_addresses(
    mnemonic: String,
    passphrase: Option<String>,
    count: Option<u32>,
) -> Result<Vec<signing::MnemonicAddressEntry>, String> {
    dev_secrets::ensure_dev_mnemonic_signing_allowed()?;
    signing::list_mnemonic_addresses(
        &mnemonic,
        passphrase.as_deref().unwrap_or(""),
        count.unwrap_or(20),
    )
}

#[tauri::command]
pub(crate) fn sign_message_with_mnemonic_path(
    mnemonic: String,
    passphrase: Option<String>,
    derivation_path: String,
    message: String,
) -> Result<signing::SignatureResult, String> {
    dev_secrets::ensure_dev_mnemonic_signing_allowed()?;
    signing::sign_message_with_mnemonic_path(
        &mnemonic,
        passphrase.as_deref().unwrap_or(""),
        &derivation_path,
        &message,
    )
}

#[tauri::command]
pub(crate) fn sign_with_mnemonic_path(
    mnemonic: String,
    passphrase: Option<String>,
    derivation_path: String,
    sighash_hex: String,
) -> Result<signing::SignatureResult, String> {
    dev_secrets::ensure_dev_mnemonic_signing_allowed()?;
    signing::sign_with_mnemonic_path(
        &mnemonic,
        passphrase.as_deref().unwrap_or(""),
        &derivation_path,
        &sighash_hex,
    )
}

File: desktop-app/src/api/signing.ts
import { z } from 'zod'
import type { ApiResult } from '@/types'
import { tauriCall } from './tauri-bridge'
import { decodedActionSchema, sighashResultSchema, signatureResultSchema, verifyResultSchema } from './ipc-schemas'

export type SighashResult = {
	sighashHex: string
	seqno: number
}

export type VerifyResult = {
	valid: boolean
	signaturesVerified: number
	thresholdRequired: number
}

export type SignatureResult = {
	publicKeyHex: string
	signatureHex: string
}

export type DecodedAction =
	| { kind: 'multisig_update'; role: string; addKeys: string[]; removeKeys: string[]; newThreshold: number }
	| { kind: 'vk_update'; authority: string; typeId: number; conditionHex: string }
	| { kind: 'defcon_1' }
	| { kind: 'defcon_3' }
	| { kind: 'unknown'; rawHex: string }

export function decodeActionHex(actionHex: string): Promise<ApiResult<DecodedAction>> {
	return tauriCall<DecodedAction>('decode_action_hex', { actionHex }, decodedActionSchema)
}

export function computeSighash(seqno: number, actionHex: string): Promise<ApiResult<SighashResult>> {
	return tauriCall('compute_sighash', { seqno, actionHex }, sighashResultSchema)
}

/**
 * Canonical SPS-65 signing message (the exact text the device signs) for an action.
 * Used to show what the connected device displays: Trezor renders this text; Ledger
 * renders its SHA-256 ("Message hash"). The BIP-137 sighash never appears on-device.
 */
export function renderSigningMessage(seqno: number, actionHex: string): Promise<ApiResult<string>> {
	return tauriCall('render_signing_message', { seqno, actionHex }, z.string())
}

export function verifyThreshold(
	publicKeysHex: string[],
	threshold: number,
	signaturesHex: string[],
	sighashHex: string,
): Promise<ApiResult<VerifyResult>> {
	return tauriCall('verify_threshold', { publicKeysHex, threshold, signaturesHex, sighashHex }, verifyResultSchema)
}

export function signSighashMock(secretKeyHex: string, sighashHex: string): Promise<ApiResult<SignatureResult>> {
	return tauriCall('sign_sighash_mock', { secretKeyHex, sighashHex }, signatureResultSchema)
}
